common/src/subgraph_client/client.rs

```diff
@@ -0,0 +1,202 @@
+// Copyright 2023-, GraphOps and Semiotic Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::anyhow;
+use eventuals::Eventual;
+use graphql::http::Response;
+use log::warn;
+use reqwest::{header, Url};
+use serde::de::Deserialize;
+use serde_json::Value;
+use toolshed::thegraph::DeploymentId;
+
+use super::monitor::{monitor_deployment_status, DeploymentStatus};
+
+#[derive(Debug, Clone)]
+pub struct DeploymentDetails {
+    pub deployment: Option<DeploymentId>,
+    pub status_url: Option<Url>,
+    pub query_url: Url,
+}
+
+impl DeploymentDetails {
+    pub fn for_graph_node(
+        graph_node_base_url: &str,
+        deployment: DeploymentId,
+    ) -> Result<Self, anyhow::Error> {
+        Ok(Self {
+            deployment: Some(deployment),
+            status_url: Some(Url::parse(&format!("{graph_node_base_url}/status"))?),
+            query_url: Url::parse(&format!("{graph_node_base_url}/subgraphs/id/{deployment}"))?,
+        })
+    }
+
+    pub fn for_query_url(query_url: &str) -> Result<Self, anyhow::Error> {
+        Ok(Self {
+            deployment: None,
+            status_url: None,
+            query_url: Url::parse(query_url)?,
+        })
+    }
+}
+
+struct DeploymentClient {
+    pub status: Option<Eventual<DeploymentStatus>>,
+    pub query_url: Url,
+}
+
+impl DeploymentClient {
+    pub fn new(details: DeploymentDetails) -> Self {
+        Self {
+            status: details
+                .deployment
+                .zip(details.status_url)
+                .map(|(deployment, url)| monitor_deployment_status(deployment, url)),
+            query_url: details.query_url,
+        }
+    }
+
+    pub async fn query<T: for<'de> Deserialize<'de>>(
+        &self,
+        body: &Value,
+    ) -> Result<Response<T>, anyhow::Error> {
+        if let Some(ref status) = self.status {
+            let deployment_status = status.value().await.expect("reading deployment status");
+
+            if !deployment_status.synced || &deployment_status.health != "healthy" {
+                return Err(anyhow!(
+                    "Deployment `{}` is not ready or healthy to be queried",
+                    self.query_url
+                ));
+            }
+        }
+
+        Ok(reqwest::Client::new()
+            .post(self.query_url.as_ref())
+            .json(body)
+            .header(header::USER_AGENT, "indexer-common")
+            .header(header::CONTENT_TYPE, "application/json")
+            .send()
+            .await
+            .and_then(|response| response.error_for_status())?
+            .json::<Response<T>>()
+            .await?)
+    }
+}
+
+/// Network subgraph query wrapper
+///
+/// This is Arc internally, so it can be cloned and shared between threads.
```

Comment:
I think it's not strictly the case anymore.

Reply:
Addressed!